pkg/kube/config.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// BuildClientConfig is a helper function that builds client config from a kubeconfig filepath.
+//
+// This is a modified version of k8s.io/client-go/tools/clientcmd/BuildConfigFromFlags with the
+// difference that it loads default configs if not running in-cluster and no explicit path.
+func BuildClientConfig(kubeconfig string) (*rest.Config, error) {
+	if kubeconfig == "" {
+		// Try using the inClusterConfig
+		config, err := rest.InClusterConfig()
+		if err == nil {
+			// Using In-Cluster client config
+			return config, nil
+		}
+	}
+
+	//Config loading rules:
+	// 1. kubeconfig if it not empty string
+	// 2. Config(s) in KUBECONFIG environment variable
+	// 3. Use $HOME/.kube/config
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	loadingRules.ExplicitPath = kubeconfig
+	configOverrides := &clientcmd.ConfigOverrides{}
+	return clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides).ClientConfig()
```

Comment:
On 2nd pass, it looks like the logic here is inverted. We should check kubeconfig, envvars, etc. first and then fallback to in-cluster config. As-is, `istioctl` will try in-cluster config _first_ unless user explicitly sets a kubeconfig (valid or not). 

For reference, kubectl logic for setting kubeconfig is [here](https://github.com/kubernetes/kubernetes/blob/86ae84b10ea26f5bfe39a297c9f4fc386551aa5b/pkg/kubectl/genericclioptions/config_flags.go#L108) which seems to rely on NewNonInteractiveDeferredLoadingClientConfig handling the in-cluster case.

Reply:
Thanks @ayj 
Indeed when using the DefaultClientConfig it will also handle in-cluster so no need to explicitly handle it.
I updated the code and tested the following scenarios:
1. In-cluster with "" kubeconfig would load the in-cluster config
1. In-cluster with "any" (invalid or empty) kubeconfig would load the in-cluster config
1. Out-of-cluster with "" kubeconfig would load the default config
1. Out-of-cluster with valid "any" kubeconfig would load the config from the specified file
1. Out-of-cluster with invalid "any" kubeconfig would load the default config

I assumed a fallback to default in any case where specified `kubeconfig` is invalid (e.g. doesn't exists). Do you think that in such cases we want to return an error instead?